src/utils/function.js

```diff
@@ -44,11 +44,50 @@ export function rateLimit(win, callback, minInterval) {
     }
   }
 
-  return function() {
+  return function(...args) {
     if (locker) {
-      nextCallArgs = arguments;
+      nextCallArgs = args;
     } else {
-      fire(arguments);
+      fire(args);
+    }
+  };
+}
+
+/**
+ * Wraps a given callback and applies a wait timer, so that minInterval
+ * milliseconds must pass since the last call before the callback is actually
+ * invoked.
+ *
+ * @param {!Window} win
+ * @param {function()} callback
+ * @param {number} minInterval the minimum time interval in millisecond
+ * @returns {function()}
+ */
+export function debounce(win, callback, minInterval) {
+  let locker = 0;
+  let timestamp = 0;
+  let nextCallArgs = null;
+
+  function fire(args) {
+    nextCallArgs = null;
+    callback.apply(null, args);
+  }
+
+  function waiter() {
+    locker = 0;
+    const remaining = minInterval - (Date.now() - timestamp);
+    if (remaining > 0) {
+      locker = win.setTimeout(waiter, remaining);
+    } else {
+      fire(nextCallArgs);
+    }
+  }
+
+  return function(...args) {
+    timestamp = Date.now();
```

Comment:
I am worried that this API only works nicely when used directly as a `listener` but becomes fragile in other cases.
If we take the PositionObserver use-case, I assume it would becomes something like:
```
addEventListener('scroll', onScroll);

const debouncedScroll = debounce(onEnd, 500);
function onScroll() {
  isScrolling = true;
  debouncedScroll();
}

function onEnd() {
  isScrolling = false;
}
```

This looks a bit hard to read (what is even a clean name for variable that holds the return of `debounce`?) also a bit fragile (e.g. replacing `debouncedScroll();` with `debounce(onEnd, 500);` will break the code).

Why not something like:
```
addEventListener('scroll', onScroll);

listenerTimeout(onScroll, 500).then(onEnd);

function onScroll() {
  isScrolling = true;
}

function onEnd() {
  isScrolling = false;
}
```

does this make sense or am I off base here? ( I do understand current API is common in JS libraries and works nicely for cases like `addEventListener('scroll', debounce(onEnd, 500));`)